vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java

```diff
@@ -109,4 +117,228 @@ public void treegrid_select_singleItemSignature_selectFromClient() {
+    }
+
+    @Test
+    public void selectFromClient_lazyDefinedSizeDataProviderWithHidden_skipsUpdateCheckboxState() {
+        verifyUpdateSelectAllCheckboxStateWhenSelectFromClientInMultiSelectMode(
+                false, false, false, false,
+                GridMultiSelectionModel.SelectAllCheckboxVisibility.HIDDEN);
+
+    }
+
+    @Test
+    public void selectFromClient_lazyUnknownSizeDataProviderWithHidden_skipsUpdateCheckboxState() {
+        verifyUpdateSelectAllCheckboxStateWhenSelectFromClientInMultiSelectMode(
+                false, true, false, false,
+                GridMultiSelectionModel.SelectAllCheckboxVisibility.HIDDEN);
+
+    }
+
+    private void verifySelectAllCheckboxVisibilityInMultiSelectMode(
+            boolean inMemory, boolean unknownItemCount,
+            boolean expectedVisibility,
+            GridMultiSelectionModel.SelectAllCheckboxVisibility visibility) {
+        grid.setSelectionMode(SelectionMode.MULTI);
+
+        GridMultiSelectionModel<String> selectionModel = (GridMultiSelectionModel<String>) grid
+                .getSelectionModel();
+        selectionModel.setSelectAllCheckboxVisibility(visibility);
+
+        DataProvider<String, ?> dataProvider;
+
+        if (inMemory) {
+            dataProvider = getInMemoryDataProvider();
+        } else if (unknownItemCount) {
+            dataProvider = getUnknownItemCountLazyDataProvider();
+        } else {
+            dataProvider = getDefinedSizeLazyDataProvider();
+        }
+
+        grid.setDataProvider(dataProvider);
+        grid.getDataCommunicator().setDefinedSize(!unknownItemCount);
+
+        boolean selectAllCheckboxVisible = selectionModel
+                .isSelectAllCheckboxVisible();
+
+        Assert.assertEquals(
+                "Unexpected select all checkbox visibility in multi-select mode",
+                expectedVisibility, selectAllCheckboxVisible);
+    }
+
+    private void verifyUpdateSelectAllCheckboxStateWhenSelectFromClientInMultiSelectMode(
+            boolean inMemory, boolean unknownItemCount,
+            boolean expectedSizeQuery, boolean expectedCheckboxStateUpdate,
+            GridMultiSelectionModel.SelectAllCheckboxVisibility visibility) {
+        grid.setSelectionMode(SelectionMode.MULTI);
+
+        GridMultiSelectionModel<String> selectionModel = (GridMultiSelectionModel<String>) grid
+                .getSelectionModel();
+        selectionModel.setSelectAllCheckboxVisibility(visibility);
+
+        DataProvider<String, ?> dataProvider;
+
+        if (inMemory) {
+            dataProvider = getInMemoryDataProvider();
+        } else if (unknownItemCount) {
+            dataProvider = getUnknownItemCountLazyDataProvider();
+        } else {
+            dataProvider = getDefinedSizeLazyDataProvider();
+        }
+
+        dataProvider = Mockito.spy(dataProvider);
+        grid.setDataProvider(dataProvider);
+        grid.getDataCommunicator().setDefinedSize(!unknownItemCount);
```

Comment:
Just noting that this is c/p same code from the above method so maybe don't duplicate it ?

Reply:
Refactored